core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java

```diff
@@ -0,0 +1,438 @@
+    final boolean fastMergeEnabled =
+      sparkConf.getBoolean("spark.shuffle.unsafe.fastMergeEnabled", true);
+    final boolean fastMergeIsSupported =
+      !compressionEnabled || compressionCodec instanceof LZFCompressionCodec;
+    try {
+      if (spills.length == 0) {
+        new FileOutputStream(outputFile).close(); // Create an empty file
+        return new long[partitioner.numPartitions()];
+      } else if (spills.length == 1) {
+        // Here, we don't need to perform any metrics updates because the bytes written to this
+        // output file would have already been counted as shuffle bytes written.
+        Files.move(spills[0].file, outputFile);
+        return spills[0].partitionLengths;
+      } else {
+        final long[] partitionLengths;
+        // There are multiple spills to merge, so none of these spill files' lengths were counted
+        // towards our shuffle write count or shuffle write time. If we use the slow merge path,
+        // then the final output file's size won't necessarily be equal to the sum of the spill
+        // files' sizes. To guard against this case, we look at the output file's actual size when
+        // computing shuffle bytes written.
+        //
+        // We allow the individual merge methods to report their own IO times since different merge
+        // strategies use different IO techniques.  We count IO during merge towards the shuffle
+        // shuffle write time, which appears to be consistent with the "not bypassing merge-sort"
+        // branch in ExternalSorter.
+        if (fastMergeEnabled && fastMergeIsSupported) {
+          // Compression is disabled or we are using an IO compression codec that supports
+          // decompression of concatenated compressed streams, so we can perform a fast spill merge
+          // that doesn't need to interpret the spilled bytes.
+          if (transferToEnabled) {
+            logger.debug("Using transferTo-based fast merge");
+            partitionLengths = mergeSpillsWithTransferTo(spills, outputFile);
+          } else {
+            logger.debug("Using fileStream-based fast merge");
+            partitionLengths = mergeSpillsWithFileStream(spills, outputFile, null);
+          }
+        } else {
+          logger.debug("Using slow merge");
+          partitionLengths = mergeSpillsWithFileStream(spills, outputFile, compressionCodec);
+        }
+        // When closing an UnsafeShuffleExternalSorter that has already spilled once but also has
+        // in-memory records, we write out the in-memory records to a file but do not count that
+        // final write as bytes spilled (instead, it's accounted as shuffle write). The merge needs
+        // to be counted as shuffle write, but this will lead to double-counting of the final
+        // SpillInfo's bytes.
+        writeMetrics.decShuffleBytesWritten(spills[spills.length - 1].file.length());
```

Comment:
If control reaches this point, then we have written multiple files to disk in our external sorter.  The last file should not be counted as bytes spilled in order to try to remain consistent with other code.  In the existing sort-merge code, the final in-memory partition is merged against the on-disk partitions and thus does not get counted as extra disk write.  Here, we end up dumping the in-memory partition to disk, incurring a bit of extra write but massively simplifying the merge path code.  It's true that we're performing some disk writes that aren't accounted for anywhere in the final metrics, but I'm not sure how this should be reported / handled. I don't think that it's correct to treat these bytes as spilled bytes, but treating them as shuffle write is also confusing because then we'll end up reporting more shuffle bytes written than are transferred over the network.

The reason that I decrement the count here is so that the final count of shuffle bytes written is equal to the output file's size.  Alternatively, I could have avoided incrementing shuffleBytesWritten when writing the last file, but that would cause complications when updating record counts.

I'm happy to work on fixes for the metrics, but I'd rather defer that to a separate followup patch that does a broader refactoring of our metrics code.